crates/engine/tree/src/tree/payload_validator.rs

```diff
@@ -374,105 +389,35 @@ where
 
-        // We only run the parallel state root if we are not currently persisting any blocks or
-        // persisting blocks that are all ancestors of the one we are executing.
-        //
-        // If we're committing ancestor blocks, then: any trie updates being committed are a subset
-        // of the in-memory trie updates collected before fetching reverts. So any diff in
-        // reverts (pre vs post commit) is already covered by the in-memory trie updates we
-        // collect in `compute_state_root_parallel`.
-        //
-        // See https://github.com/paradigmxyz/reth/issues/12688 for more details
-        let persisting_kind = ctx.persisting_kind_for(input.block_with_parent());
-        // don't run parallel if state root fallback is set
-        let run_parallel_state_root =
-            persisting_kind.can_run_parallel_state_root() && !self.config.state_root_fallback();
-
-        // Use state root task only if:
-        // 1. No persistence is in progress
-        // 2. Config allows it
-        // 3. No ancestors with missing trie updates. If any exist, it will mean that every state
-        //    root task proof calculation will include a lot of unrelated paths in the prefix sets.
-        //    It's cheaper to run a parallel state root that does one walk over trie tables while
-        //    accounting for the prefix sets.
+        // Plan the strategy used for state root computation.
+        let state_root_plan = self.plan_state_root_computation(&input, &ctx);
+        let persisting_kind = state_root_plan.persisting_kind;
         let has_ancestors_with_missing_trie_updates =
-            self.has_ancestors_with_missing_trie_updates(input.block_with_parent(), ctx.state());
-        let mut use_state_root_task = run_parallel_state_root &&
-            self.config.use_state_root_task() &&
-            !has_ancestors_with_missing_trie_updates;
+            state_root_plan.has_ancestors_with_missing_trie_updates;
+        let mut strategy = state_root_plan.strategy;
 
         debug!(
             target: "engine::tree",
             block=?block_num_hash,
-            run_parallel_state_root,
-            has_ancestors_with_missing_trie_updates,
-            use_state_root_task,
-            config_allows_state_root_task=self.config.use_state_root_task(),
+            ?strategy,
+            ?has_ancestors_with_missing_trie_updates,
             "Deciding which state root algorithm to run"
         );
 
         // use prewarming background task
         let txs = self.tx_iterator_for(&input)?;
-        let mut handle = if use_state_root_task {
-            // use background tasks for state root calc
-            let consistent_view =
-                ensure_ok!(ConsistentDbView::new_with_latest_tip(self.provider.clone()));
-
-            // get allocated trie input if it exists
-            let allocated_trie_input = self.payload_processor.take_trie_input();
```

Comment:
the logic here is convoluted, so would need some sanity check here @mediocregopher @Rjected @shekhirin to see if the simplified logic below matches